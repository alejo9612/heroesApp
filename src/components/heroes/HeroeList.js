import React,{useMemo} from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroeCard } from './HeroeCard';

export const HeroeList = ({publisher}) => {

    //esto lo que hace es que me memorizará la función que necesito solo basandose de la función ya descrita y comenta
    const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher]);

    //const heroes = getHeroesByPublisher(publisher);

    return (
        <div className="row animate__animated animate__fadeIn">
            {
                heroes.map(heroe => (
                    <HeroeCard key={heroe.id}   
                        {...heroe}
                    >
                    </HeroeCard>
                ))
            }
        </div>
    )
}
