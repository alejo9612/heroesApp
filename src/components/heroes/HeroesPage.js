import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router';
import { getHeroesById } from '../../selectors/getHeroesById';
//import batman from '../../assets/heroes/dc-batman';

//const heroImagenes = require.context('../../assets/heroes',true); declaración de weepack

export const HeroesPage = ({history}) => {

    const {heroeId} = useParams();

    const heroe = useMemo(() => getHeroesById(heroeId), [heroeId])
    //const heroe = getHeroesById(heroeId);

    if (!heroe) {
        return <Redirect to="/"/>
    }

    const handleReturn = () =>{

        if (history.length <=2) {
            history.push('/');
        }else{
            history.goBack(); //me redirecciona al menu en el que este
        }
    }

    const {superhero, alter_ego, first_appearance, characters, publisher} = heroe;

    // console.log(history);
    
    return (
        <div className="row mt-5">
            <div className="col-4">
                <img src={`../assets/heroes/${heroeId}.jpg`} //con la imagenes en la carpeta Public assets
                     //src={batman} con la importacción de batman del asset dentro de src
                     //src={heroImagenes(`./${heroeId}.jpg`)} //importación propia del weepack
                     alt={superhero}
                     className="img-thumbnail animate__animated animate__fadeInUp"
                />
            </div>
            <div className="col-8">
                <h5>{superhero}</h5>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><b>Alter Ego: </b>{alter_ego}</li>
                    <li className="list-group-item"><b>Publisher: </b>{publisher}</li>
                    <li className="list-group-item"><b>First Appearance: </b>{first_appearance}</li>
                </ul>
                <h5>Characters</h5>
                <p>{characters}</p>

                <button className="btn btn-outline-primary"
                        onClick={handleReturn}
                >
                    Regresar
                </button>
            </div>
            
        </div>
    )
}
