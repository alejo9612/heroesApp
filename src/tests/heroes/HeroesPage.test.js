import React from "react";
import { mount } from "enzyme";
import { HeroesPage } from "../../components/heroes/HeroesPage";
import { MemoryRouter, Route } from 'react-router-dom';

describe('Pruebas en el <HeroesPage/>', () => {

    //simulamo el hisory con sus funcines y demás
    const history = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }

    const wrapper = mount(
        <MemoryRouter initialEntries={['/hero']}>
            <HeroesPage history={history} />
        </MemoryRouter>
    );

    test('Debe de mostrar el componenete con la redirección  en caso de no haber argumentos en la URL ', () => {

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);

    });

    test('Debe de mostrar un heroe si el parametro exite y se encuentra', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
                {/* validamos los parametros */}
                <Route path="/heroe/:heroeId" component={HeroesPage} />
            </MemoryRouter>
        );
        expect(wrapper.find('.row').exists()).toBe(true);
    });

    test('Debe de regresar a la pantalla anterior con el PUSH', () => {
        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        //* validamos los parametros */}
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
                <Route path="/heroe/:heroeId"
                    component={() => <HeroesPage history={history} />}
                />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();


        expect(history.push).toHaveBeenCalledWith('/');
        expect(history.goBack).not.toHaveBeenCalledWith();
    });

    test('Debe de regresar a la pantalla anterior', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
                {/* validamos los parametros */}
                <Route path="/heroe/:heroeId"
                    component={(props) => <HeroesPage history={history} />} />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        expect(history.push).toHaveBeenCalledTimes(0)
        expect(history.goBack).toHaveBeenCalledWith();
    });

});

