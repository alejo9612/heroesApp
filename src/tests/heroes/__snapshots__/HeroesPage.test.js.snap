// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <HeroesPage/> Debe de mostrar el componenete con la redirecci√≥n  en caso de no haber argumentos en la URL  1`] = `""`;
